torch/nn/modules/module.py

```diff
@@ -2046,16 +2046,9 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
 
                 try:
                     with torch.no_grad():
-                        if assign_to_params_buffers:
-                            # Shape checks are already done above
-                            if (isinstance(param, torch.nn.Parameter) and
-                                    not isinstance(input_param, torch.nn.Parameter)):
-                                setattr(self, name, torch.nn.Parameter(input_param, requires_grad=param.requires_grad))
-                            else:
-                                setattr(self, name, input_param)
-                        elif use_swap_tensors:
+                        if use_swap_tensors:
                             param_requires_grad = param.requires_grad
-                            new_input_param = param.module_load(input_param)
+                            new_input_param = param.module_load(input_param, assign=assign_to_params_buffers)
```

Comment:
should we add an assert to make sure that `__torch_function__` overrides conform to the definition for `assign` or should we let the torch function handler do anything?

Reply:
How would you check that it conforms to it?

Reply:
Something like 

```
if assign:
         assert type(new_input_param) == type(input_param)
else:
         assert type(new_input_param) == type(param)
 ```
 
 But perhaps this is too restrictive

Reply:
This feels too restrictive indeed, We did mention that in the subclass case, even when assign=False, we do say that you can change the type there.